File: android/src/main/java/com/capacitorjs/plugins/googlemaps/GoogleMapsPreferences.kt
package com.capacitorjs.plugins.googlemaps

import com.google.android.gms.maps.model.LatLng
import org.json.JSONObject

class GoogleMapsPreferences(fromJSONObject: JSONObject) {
        var zoom: GoogleMapZoomOptions? = null
        var padding: GoogleMapPadding? = null
        var building: Boolean? = null
        var gestureBounds: Array<LatLng>? = null

    init {
        if (fromJSONObject.has("zoom")) {
            val zoomObj = fromJSONObject.getJSONObject("zoom")
            zoom = GoogleMapZoomOptions(zoomObj)
        }
        if (fromJSONObject.has("padding")) {
            val paddingObj = fromJSONObject.getJSONObject("padding")
            padding = GoogleMapPadding(paddingObj)
        }
        if (fromJSONObject.has("building")) {
            building = fromJSONObject.getBoolean("building")
        }
        if (fromJSONObject.has("gestureBounds")) {
            val gestureBoundsJSONArray = fromJSONObject.getJSONArray("gestureBounds")
            val list = mutableListOf<LatLng>()
            for (i in 0 until gestureBoundsJSONArray.length()) {
                val gestureBoundObj = gestureBoundsJSONArray.getJSONObject(i)
                if (!gestureBoundObj.has("lat") || !gestureBoundObj.has("lng")) {
                    throw InvalidArgumentsError("One of the LatLng objects is missing 'lat' and/or 'lng'")
                }
                val lat = gestureBoundObj.getDouble("lat")
                val lng = gestureBoundObj.getDouble("lng")
                list.add(LatLng(lat, lng))
            }
            gestureBounds = list.toTypedArray()
        }
    }
}

File: ios/Plugin/GoogleMapsPreferences.swift
import Foundation
import Capacitor

public struct GoogleMapsPreferences: Codable {
        let zoom: GoogleMapZoomOptions?
        let padding: GoogleMapPadding?
        let building: Bool?
        let gestureBounds: [LatLng]?
    
    init(fromJSObject: JSObject) throws {
        building = fromJSObject["building"] as? Bool
        
        if let zoomObj = fromJSObject["zoom"] as? JSObject {
            let minZoom = zoomObj["minZoom"] as? Float
            let maxZoom = zoomObj["maxZoom"] as? Float
            self.zoom = GoogleMapZoomOptions(minZoom:minZoom, maxZoom:maxZoom)
        } else {
            self.zoom = nil
        }
        
        if let paddingObj = fromJSObject["padding"] as? JSObject {
            self.padding = try? GoogleMapPadding(fromJSObject: paddingObj)
        } else {
            self.padding = nil
        }
        
        if let gestureBoundsArray = fromJSObject["gestureBounds"] as? [JSObject] {
             var parsedGestureBounds: [LatLng] = []
             for obj in gestureBoundsArray {
                 guard let lat = obj["lat"] as? Double,
                       let lng = obj["lng"] as? Double else {
                     throw GoogleMapErrors.invalidArguments("Each gestureBound in 'gestureBounds' must have 'lat' and 'lng'")
                 }
                 parsedGestureBounds.append(LatLng(lat: lat, lng: lng))
             }
             self.gestureBounds = parsedGestureBounds
         } else {
             self.gestureBounds = nil
         }

    }
}
